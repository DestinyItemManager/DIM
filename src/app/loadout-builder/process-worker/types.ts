import { ArmorStatHashes, ArmorStats, LockableBucketHash, StatRanges } from '../types';

export interface ProcessResult {
  /** A small number of the best sets, depending on operation mode. */
  sets: ProcessArmorSet[];
  /** The total number of combinations considered. */
  combos: number;
  /** The stat ranges of all sets that matched our filters & mod selection. */
  statRangesFiltered?: StatRanges;
  /** Statistics about how many sets passed/failed the constraints, for error reporting */
  processInfo?: ProcessStatistics;
}

export interface ProcessItem {
  id: string;
  hash: number;
  name: string;
  isExotic: boolean;
  isArtifice: boolean;
  /** The remaining assumed energy capacity for this item, after assigning slot-specific mods */
  remainingEnergyCapacity: number;
  power: number;
  stats: { [statHash: number]: number };
  compatibleModSeasons?: string[];
}

export type ProcessItemsByBucket = {
  [bucketHash in LockableBucketHash]: ProcessItem[];
};

export interface ProcessArmorSet {
  /** The overall stats for the loadout as a whole, including subclass, mods and including auto stat mods. */
  readonly stats: Readonly<ArmorStats>;
  /** The assumed stats from the armor items themselves only. */
  readonly armorStats: Readonly<ArmorStats>;
  /** For each armor type (see LockableBuckets), this is the list of items that could interchangeably be put into this loadout. */
  readonly armor: readonly string[];
  /** Which stat mods were added? */
  readonly statMods: number[];
}

export interface IntermediateProcessArmorSet {
  /** The overall stats for the loadout as a whole, EXCLUDING auto stat mods. */
  stats: number[];
  /** The first (highest-power) valid set from this stat mix. */
  armor: ProcessItem[];
}

export interface ProcessMod {
  hash: number;
  /** The energy cost of the mod. */
  energyCost: number;
  /** This should only be available in legacy, combat and raid mods */
  tag?: string;
}

/**
 * Data describing the mods that can be automatically picked.
 */
export interface AutoModData {
  generalMods: {
    [key in ArmorStatHashes]?: {
      majorMod: { hash: number; cost: number };
      minorMod: { hash: number; cost: number };
    };
  };
  artificeMods: { [key in ArmorStatHashes]?: { hash: number } };
}

export interface LockedProcessMods {
  generalMods: ProcessMod[];
  activityMods: ProcessMod[];
}

export interface RejectionRate {
  timesFailed: number;
  timesChecked: number;
}

export interface ModAssignmentStatistics {
  /** Mod-tag and mod element counts check. */
  earlyModsCheck: RejectionRate;
  /** How many times we couldn't possibly hit the target stats with any number of auto mods picks */
  autoModsPick: RejectionRate;
  finalAssignment: {
    /** How many times we tried mod permutations for permutations that worked. */
    modAssignmentAttempted: number;
    /** How many times we failed to assign user-picked slot-independent mods. */
    modsAssignmentFailed: number;
    /** How many times we failed to assign auto stat mods. */
    autoModsAssignmentFailed: number;
  };
}

/**
 * Information about the operation of the worker process.
 */
export interface ProcessStatistics {
  numProcessed: number;
  numValidSets: number;
  statistics: {
    /** Sets skipped for really uninteresting/coarse reasons. */
    skipReasons: {
      noExotic: number;
      doubleExotic: number;
      skippedLowTier: number;
    };
    lowerBoundsExceeded: RejectionRate;
    upperBoundsExceeded: RejectionRate;
    modsStatistics: ModAssignmentStatistics;
  };
}
