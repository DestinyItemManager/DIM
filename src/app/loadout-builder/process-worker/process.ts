import _ from 'lodash';
import { infoLog } from '../../utils/log';
import {
  ArmorStatHashes,
  ArmorStats,
  LockableBucketHashes,
  LockableBuckets,
  StatFilters,
  StatRanges,
} from '../types';
import { pickAndAssignSlotIndependentMods, precalculateStructures } from './process-utils';
import { SetTracker } from './set-tracker';
import {
  LockedProcessMods,
  ProcessArmorSet,
  ProcessItem,
  ProcessItemsByBucket,
  ProcessStatistics,
} from './types';

/** Caps the maximum number of total armor sets that'll be returned */
const RETURNED_ARMOR_SETS = 200;

/**
 * This processes all permutations of armor to build sets
 * @param filteredItems pared down list of items to process sets from
 * @param modStatTotals Stats that are applied to final stat totals, think general and other mod stats
 */
export function process(
  filteredItems: ProcessItemsByBucket,
  /** Selected mods' total contribution to each stat */
  modStatTotals: ArmorStats,
  /** Mods to add onto the sets */
  lockedMods: LockedProcessMods,
  /** The user's chosen stat order, including disabled stats */
  statOrder: ArmorStatHashes[],
  statFilters: StatFilters,
  /** Ensure every set includes one exotic */
  anyExotic: boolean,
  /** Use stat mods to hit stat minimums */
  autoStatMods: boolean,
  onProgress: (remainingTime: number) => void
): {
  sets: ProcessArmorSet[];
  combos: number;
  /** The stat ranges of all sets that matched our filters & mod selection. */
  statRangesFiltered?: StatRanges;
  processInfo?: ProcessStatistics;
} {
  const pstart = performance.now();

  const modStatsInStatOrder = statOrder.map((h) => modStatTotals[h]);
  const statFiltersInStatOrder = statOrder.map((h) => statFilters[h]);

  // This stores the computed min and max value for each stat as we process all sets, so we
  // can display it on the stat filter dropdowns
  const statRangesFiltered: StatRanges = _.mapValues(statFilters, () => ({
    min: 100,
    max: 0,
  }));
  const statRangesFilteredInStatOrder = statOrder.map((h) => statRangesFiltered[h]);

  // Store stat arrays for each items in stat order
  const statsCacheInStatOrder = new Map<ProcessItem, number[]>();

  // Precompute the stats of each item in stat order
  for (const item of LockableBucketHashes.flatMap((h) => filteredItems[h])) {
    statsCacheInStatOrder.set(
      item,
      statOrder.map((statHash) => Math.max(item.stats[statHash], 0))
    );
  }

  // Each of these groups has already been reduced (in useProcess.ts) to the
  // minimum number of examples that are worth considering.
  const helms = filteredItems[LockableBuckets.helmet];
  const gauntlets = filteredItems[LockableBuckets.gauntlets];
  const chests = filteredItems[LockableBuckets.chest];
  const legs = filteredItems[LockableBuckets.leg];
  const classItems = filteredItems[LockableBuckets.classitem];

  // The maximum possible combos we could have
  const combos = helms.length * gauntlets.length * chests.length * legs.length * classItems.length;
  const numItems =
    helms.length + gauntlets.length + chests.length + legs.length + classItems.length;

  infoLog('loadout optimizer', 'Processing', combos, 'combinations from', numItems, 'items', {
    helms: helms.length,
    gauntlets: gauntlets.length,
    chests: chests.length,
    legs: legs.length,
    classItems: classItems.length,
  });

  if (combos === 0) {
    return { sets: [], combos: 0 };
  }

  const setTracker = new SetTracker(10_000);

  const { activityMods, combatMods, generalMods } = lockedMods;

  const precalculatedInfo = precalculateStructures(
    generalMods,
    combatMods,
    activityMods,
    autoStatMods,
    statOrder
  );
  const hasMods = Boolean(combatMods.length || activityMods.length || generalMods.length);

  const setStatistics = {
    skipReasons: {
      doubleExotic: 0,
      noExotic: 0,
      skippedLowTier: 0,
    },
    lowerBoundsExceeded: { timesChecked: 0, timesFailed: 0 },
    upperBoundsExceeded: { timesChecked: 0, timesFailed: 0 },
    modsStatistics: {
      earlyModsCheck: { timesChecked: 0, timesFailed: 0 },
      autoModsPick: { timesChecked: 0, timesFailed: 0 },
      finalAssignment: {
        modAssignmentAttempted: 0,
        modsAssignmentFailed: 0,
        autoModsAssignmentFailed: 0,
      },
    },
  };
  const processStatistics: ProcessStatistics = {
    numProcessed: 0,
    numValidSets: 0,
    statistics: setStatistics,
  };

  let elapsedSeconds = 0;

  for (const helm of helms) {
    for (const gaunt of gauntlets) {
      // For each additional piece, skip the whole branch if we've managed to get 2 exotics
      if (gaunt.isExotic && helm.isExotic) {
        setStatistics.skipReasons.doubleExotic += chests.length * legs.length * classItems.length;
        continue;
      }
      for (const chest of chests) {
        if (chest.isExotic && (gaunt.isExotic || helm.isExotic)) {
          setStatistics.skipReasons.doubleExotic += legs.length * classItems.length;
          continue;
        }
        for (const leg of legs) {
          if (leg.isExotic && (chest.isExotic || gaunt.isExotic || helm.isExotic)) {
            setStatistics.skipReasons.doubleExotic += classItems.length;
            continue;
          }

          if (anyExotic && !helm.isExotic && !gaunt.isExotic && !chest.isExotic && !leg.isExotic) {
            setStatistics.skipReasons.noExotic += classItems.length;
            continue;
          }

          for (const classItem of classItems) {
            processStatistics.numProcessed++;

            const helmStats = statsCacheInStatOrder.get(helm)!;
            const gauntStats = statsCacheInStatOrder.get(gaunt)!;
            const chestStats = statsCacheInStatOrder.get(chest)!;
            const legStats = statsCacheInStatOrder.get(leg)!;
            const classItemStats = statsCacheInStatOrder.get(classItem)!;

            // JavaScript engines apparently don't unroll loops automatically and this makes a big difference in speed.
            const stats = [
              modStatsInStatOrder[0] +
                helmStats[0] +
                gauntStats[0] +
                chestStats[0] +
                legStats[0] +
                classItemStats[0],
              modStatsInStatOrder[1] +
                helmStats[1] +
                gauntStats[1] +
                chestStats[1] +
                legStats[1] +
                classItemStats[1],
              modStatsInStatOrder[2] +
                helmStats[2] +
                gauntStats[2] +
                chestStats[2] +
                legStats[2] +
                classItemStats[2],
              modStatsInStatOrder[3] +
                helmStats[3] +
                gauntStats[3] +
                chestStats[3] +
                legStats[3] +
                classItemStats[3],
              modStatsInStatOrder[4] +
                helmStats[4] +
                gauntStats[4] +
                chestStats[4] +
                legStats[4] +
                classItemStats[4],
              modStatsInStatOrder[5] +
                helmStats[5] +
                gauntStats[5] +
                chestStats[5] +
                legStats[5] +
                classItemStats[5],
            ];

            // TODO: avoid min/max?
            const tiers = [
              Math.min(Math.max(Math.floor(stats[0] / 10), 0), 10),
              Math.min(Math.max(Math.floor(stats[1] / 10), 0), 10),
              Math.min(Math.max(Math.floor(stats[2] / 10), 0), 10),
              Math.min(Math.max(Math.floor(stats[3] / 10), 0), 10),
              Math.min(Math.max(Math.floor(stats[4] / 10), 0), 10),
              Math.min(Math.max(Math.floor(stats[5] / 10), 0), 10),
            ];

            // Check whether the set exceeds our stat constraints
            let totalTier = 0;
            let statRangeExceeded = false;
            for (let index = 0; index < 6; index++) {
              const tier = tiers[index];
              const filter = statFiltersInStatOrder[index];
              if (!filter.ignored) {
                if (tier > filter.max) {
                  statRangeExceeded = true;
                }
                totalTier += tier;
              }
            }

            // Drop this set if it could never make it
            if (!setTracker.couldInsert(totalTier)) {
              setStatistics.skipReasons.skippedLowTier++;
              continue;
            }

            setStatistics.upperBoundsExceeded.timesChecked++;
            if (statRangeExceeded) {
              setStatistics.upperBoundsExceeded.timesFailed++;
              continue;
            }

            const armor = [helm, gaunt, chest, leg, classItem];

            const neededStats = [0, 0, 0, 0, 0, 0];
            let needSomeStats = false;

            // Check in which stats we're lacking
            for (let index = 0; index < 6; index++) {
              const value = Math.min(Math.max(stats[index], 0), 100);
              const filter = statFiltersInStatOrder[index];

              if (!filter.ignored) {
                const neededValue = filter.min * 10 - value;
                if (neededValue > 0) {
                  neededStats[index] = neededValue;
                  needSomeStats = true;
                }
              }
            }

            setStatistics.lowerBoundsExceeded.timesChecked++;
            if (needSomeStats && !autoStatMods) {
              setStatistics.lowerBoundsExceeded.timesFailed++;
              continue;
            }

            let statMods: number[] = [];
            // For armour 2 mods we ignore slot specific mods as we prefilter items based on energy requirements
            // TODO: this isn't a big part of the overall cost of the loop, but we could consider trying to slot
            // mods at every level (e.g. just helmet, just helmet+arms) and skipping this if they already fit.
            if (hasMods || needSomeStats) {
              const modsPick = pickAndAssignSlotIndependentMods(
                precalculatedInfo,
                setStatistics.modsStatistics,
                armor,
                needSomeStats ? neededStats : undefined
              );

              if (modsPick) {
                statMods = modsPick.modHashes;
              } else {
                continue;
              }
            }

            // Calculate the "tiers string" here, since most sets don't make it this far
            // A string version of the tier-level of each stat, must be lexically comparable
            // TODO: It seems like constructing and comparing tiersString would be expensive but it's less so
            // than comparing stat arrays element by element
            let tiersString = '';
            for (let index = 0; index < 6; index++) {
              const tier = tiers[index];
              // Make each stat exactly one code unit so the string compares correctly
              const filter = statFiltersInStatOrder[index];
              if (!filter.ignored) {
                // using a power of 2 (16) instead of 11 is faster
                tiersString += tier.toString(16);
              }

              // Track the stat ranges of sets that made it through all our filters
              const range = statRangesFilteredInStatOrder[index];
              const value = stats[index];
              if (value > range.max) {
                range.max = value;
              }
              if (value < range.min) {
                range.min = value;
              }
            }

            processStatistics.numValidSets++;
            setTracker.insert(totalTier, tiersString, armor, stats, statMods);
          }
        }
      }

      // Report speed every so often
      const totalTime = performance.now() - pstart;
      const newElapsedSeconds = Math.floor(totalTime / 500);

      if (newElapsedSeconds > elapsedSeconds) {
        elapsedSeconds = newElapsedSeconds;
        const speed = (processStatistics.numProcessed * 1000) / totalTime;
        const remaining = Math.round((combos - processStatistics.numProcessed) / speed);
        onProgress(remaining);
      }
    }
  }

  const finalSets = setTracker.getArmorSets(RETURNED_ARMOR_SETS);

  const totalTime = performance.now() - pstart;

  infoLog(
    'loadout optimizer',
    'found',
    processStatistics.numValidSets,
    'stat mixes after processing',
    combos,
    'stat combinations in',
    totalTime,
    'ms - ',
    Math.floor((combos * 1000) / totalTime),
    'combos/s',
    // Split into multiple objects so console.log will show them all expanded
    'sets outright skipped:',
    setStatistics.skipReasons,
    'lower and upper bounds:',
    setStatistics.lowerBoundsExceeded,
    setStatistics.upperBoundsExceeded,
    'mod assignment stats:',
    'early check:',
    setStatistics.modsStatistics.earlyModsCheck,
    'auto mods pick:',
    setStatistics.modsStatistics.autoModsPick,
    setStatistics.modsStatistics
  );
  infoLog('loadout optimizer', 'auto stat mods', {
    cacheHits: precalculatedInfo.cache.cacheHits,
    cacheMisses: precalculatedInfo.cache.cacheMisses,
    cacheSuccesses: precalculatedInfo.cache.cacheSuccesses,
  });

  const sets = finalSets.map(({ armor, stats, statMods }) => ({
    armor: armor.map((item) => item.id),
    stats: statOrder.reduce((statObj, statHash, i) => {
      statObj[statHash] = stats[i];
      return statObj;
    }, {}) as ArmorStats,
    statMods,
  }));

  return {
    sets,
    combos,
    statRangesFiltered,
    processInfo: processStatistics,
  };
}
