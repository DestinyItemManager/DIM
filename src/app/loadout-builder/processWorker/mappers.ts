import _ from 'lodash';
import { DimItem, DimSocket, DimSockets } from '../../inventory/item-types';
import {
  getModTypeTagByPlugCategoryHash,
  getSpecialtySocketMetadatas,
} from '../../utils/item-utils';
import {
  ArmorSet,
  LockedArmor2Mod,
  LockedArmor2ModMap,
  ModPickerCategories,
  statHashToType,
  StatTypes,
} from '../types';
import { ProcessArmorSet, ProcessItem, ProcessMod, ProcessSocket, ProcessSockets } from './types';

function mapDimSocketToProcessSocket(dimSocket: DimSocket): ProcessSocket {
  return {
    plug: dimSocket.plugged && {
      stats: dimSocket.plugged.stats,
      plugItemHash: dimSocket.plugged.plugDef.hash,
    },
    plugOptions: dimSocket.plugOptions.map((dimPlug) => ({
      stats: dimPlug.stats,
      plugItemHash: dimPlug.plugDef.hash,
    })),
  };
}

export function mapArmor2ModToProcessMod(mod: LockedArmor2Mod): ProcessMod {
  const processMod: ProcessMod = {
    hash: mod.modDef.hash,
    energy: mod.modDef.plug.energyCost && {
      type: mod.modDef.plug.energyCost.energyType,
      val: mod.modDef.plug.energyCost.energyCost,
    },
    investmentStats: mod.modDef.investmentStats,
  };

  if (mod.category === 'other' || mod.category === 'raid') {
    processMod.tag = getModTypeTagByPlugCategoryHash(mod.modDef.plug.plugCategoryHash);
  }

  return processMod;
}

/**
 * This sums up the total stat contributions across mods passed in. These are then applied
 * to the loadouts after all the items base values have been summed. This mimics how mods
 * effect stat values in game and allows us to do some preprocessing.
 */
export function getTotalModStatChanges(lockedArmor2Mods: LockedArmor2ModMap) {
  const totals: { [stat in StatTypes]: number } = {
    Mobility: 0,
    Recovery: 0,
    Resilience: 0,
    Intellect: 0,
    Discipline: 0,
    Strength: 0,
  };

  for (const category of Object.values(ModPickerCategories)) {
    for (const mod of lockedArmor2Mods[category]) {
      for (const stat of mod.modDef.investmentStats) {
        const statType = statHashToType[stat.statTypeHash];
        if (statType) {
          totals[statType] += stat.value;
        }
      }
    }
  }

  return totals;
}

function mapDimSocketsToProcessSockets(dimSockets: DimSockets): ProcessSockets {
  return {
    sockets: dimSockets.allSockets.map(mapDimSocketToProcessSocket),
    categories: dimSockets.categories.map((category) => ({
      categoryStyle: category.category.categoryStyle,
      sockets: category.sockets.map(mapDimSocketToProcessSocket),
    })),
  };
}

export function mapDimItemToProcessItem(
  dimItem: DimItem,
  modsForSlot: LockedArmor2Mod[]
): ProcessItem {
  const { bucket, id, type, name, equippingLabel, basePower, stats } = dimItem;

  const statMap: { [statHash: number]: number } = {};
  const baseStatMap: { [statHash: number]: number } = {};

  if (stats) {
    for (const { statHash, value, base } of stats) {
      statMap[statHash] = value;
      baseStatMap[statHash] = base;
    }
  }

  const modMetadatas = getSpecialtySocketMetadatas(dimItem);
  const costInitial =
    dimItem.energy && _.sumBy(modsForSlot, (mod) => mod.modDef.plug.energyCost?.energyCost || 0);
  return {
    bucketHash: bucket.hash,
    id,
    type,
    name,
    equippingLabel,
    basePower,
    stats: statMap,
    baseStats: baseStatMap,
    sockets: dimItem.sockets && mapDimSocketsToProcessSockets(dimItem.sockets),
    energy:
      dimItem.energy && costInitial !== null
        ? {
            type: dimItem.energy.energyType,
            val: costInitial,
          }
        : null,
    compatibleModSeasons: modMetadatas?.flatMap((m) => m.compatibleModTags),
    hasLegacyModSocket: Boolean(modMetadatas?.some((m) => m.slotTag === 'legacy')),
  };
}

export function hydrateArmorSet(
  processed: ProcessArmorSet,
  itemsById: { [id: string]: DimItem[] }
): ArmorSet {
  const armor: DimItem[][] = [];

  for (const itemId of processed.armor) {
    armor.push(itemsById[itemId]);
  }

  return {
    armor,
    statChoices: processed.statChoices,
    stats: processed.stats,
    maxPower: processed.maxPower,
  };
}
