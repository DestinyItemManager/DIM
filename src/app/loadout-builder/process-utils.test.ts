import { AssumeArmorMasterwork, LockArmorEnergyType } from '@destinyitemmanager/dim-api-types';
import { PluggableInventoryItemDefinition } from 'app/inventory/item-types';
import { DestinyEnergyType } from 'bungie-api-ts/destiny2';
import { StatHashes } from 'data/d2/generated-enums';
import _ from 'lodash';
import {
  elementalLightModHash,
  enhancedOperatorAugmentModHash,
  isArmor2Arms,
  isArmor2Chest,
  isArmor2ClassItem,
  isArmor2Helmet,
  isArmor2Legs,
  recoveryModHash,
} from 'testing/test-item-utils';
import { getTestDefinitions, getTestStores } from 'testing/test-utils';
import {
  generateProcessModPermutations,
  pickAndAssignSlotIndependentMods,
  precalculateStructures,
} from './process-worker/process-utils';
import { ModAssignmentStatistics, ProcessItem, ProcessMod } from './process-worker/types';
import { mapArmor2ModToProcessMod, mapDimItemToProcessItem } from './process/mappers';
import { ArmorStatHashes, MIN_LO_ITEM_ENERGY } from './types';

function modifyMod({
  mod,
  energyType,
  energyVal,
  tag,
}: {
  mod: ProcessMod;
  energyType?: DestinyEnergyType;
  energyVal?: number;
  tag?: string;
}) {
  const newMod = _.cloneDeep(mod);
  if (energyType !== undefined) {
    newMod.energy!.type = energyType;
  }

  if (energyVal !== undefined) {
    newMod.energy!.val = energyVal;
  }

  newMod.tag = tag;

  return newMod;
}

function modifyItem({
  item,
  energyType,
  energyVal,
  compatibleModSeasons,
}: {
  item: ProcessItem;
  energyType?: DestinyEnergyType;
  energyVal?: number;
  compatibleModSeasons?: string[];
}) {
  const newItem = _.cloneDeep(item);
  if (energyType !== undefined) {
    newItem.energy!.type = energyType;
  }

  if (energyVal !== undefined) {
    newItem.energy!.val = energyVal;
  }

  if (compatibleModSeasons !== undefined) {
    newItem.compatibleModSeasons = compatibleModSeasons;
  }

  return newItem;
}

// The tsconfig in the process worker folder messes with tests so they live outside of it.
describe('process-utils', () => {
  let generalMod: ProcessMod;
  let combatMod: ProcessMod;
  let activityMod: ProcessMod;

  let helmet: ProcessItem;
  let arms: ProcessItem;
  let chest: ProcessItem;
  let legs: ProcessItem;
  let classItem: ProcessItem;

  // use these for testing as they are reset after each test
  let items: ProcessItem[];
  let generalMods: ProcessMod[];
  let combatMods: ProcessMod[];
  let activityMods: ProcessMod[];

  const armorEnergyRules = {
    assumeArmorMasterwork: AssumeArmorMasterwork.None,
    lockArmorEnergyType: LockArmorEnergyType.All,
    minItemEnergy: MIN_LO_ITEM_ENERGY,
  };

  beforeAll(async () => {
    const [defs, stores] = await Promise.all([getTestDefinitions(), getTestStores()]);
    for (const store of stores) {
      for (const storeItem of store.items) {
        if (!helmet && isArmor2Helmet(storeItem)) {
          helmet = mapDimItemToProcessItem({
            dimItem: storeItem,
            armorEnergyRules,
          });
        }
        if (!arms && isArmor2Arms(storeItem)) {
          arms = mapDimItemToProcessItem({
            dimItem: storeItem,
            armorEnergyRules,
          });
        }
        if (!chest && isArmor2Chest(storeItem)) {
          chest = mapDimItemToProcessItem({
            dimItem: storeItem,
            armorEnergyRules,
          });
        }
        if (!legs && isArmor2Legs(storeItem)) {
          legs = mapDimItemToProcessItem({
            dimItem: storeItem,
            armorEnergyRules,
          });
        }
        if (!classItem && isArmor2ClassItem(storeItem)) {
          classItem = mapDimItemToProcessItem({
            dimItem: storeItem,
            armorEnergyRules,
          });
        }

        if (helmet && arms && chest && legs && classItem) {
          break;
        }
      }
    }

    generalMod = mapArmor2ModToProcessMod(
      defs.InventoryItem.get(recoveryModHash) as PluggableInventoryItemDefinition
    );
    combatMod = mapArmor2ModToProcessMod(
      defs.InventoryItem.get(elementalLightModHash) as PluggableInventoryItemDefinition
    );
    activityMod = mapArmor2ModToProcessMod(
      defs.InventoryItem.get(enhancedOperatorAugmentModHash) as PluggableInventoryItemDefinition
    );

    items = [helmet, arms, chest, legs, classItem];
    generalMods = [generalMod, generalMod, generalMod, generalMod, generalMod];
    combatMods = [combatMod, combatMod, combatMod, combatMod, combatMod];
    activityMods = [activityMod, activityMod, activityMod, activityMod, activityMod];
  });

  const canTakeSlotIndependentMods = (
    generalMods: ProcessMod[],
    combatMods: ProcessMod[],
    activityMods: ProcessMod[],
    items: ProcessItem[]
  ) => {
    const statOrder: ArmorStatHashes[] = [
      StatHashes.Mobility,
      StatHashes.Resilience,
      StatHashes.Recovery,
      StatHashes.Discipline,
      StatHashes.Intellect,
      StatHashes.Strength,
    ];
    const neededStats = [0, 0, 0, 0, 0, 0];
    const precalculatedInfo = precalculateStructures(
      generalMods,
      combatMods,
      activityMods,
      false,
      statOrder
    );
    const modStatistics: ModAssignmentStatistics = {
      earlyModsCheck: { timesChecked: 0, timesFailed: 0 },
      autoModsPick: { timesChecked: 0, timesFailed: 0 },
      finalAssignment: {
        modAssignmentAttempted: 0,
        modsAssignmentFailed: 0,
        autoModsAssignmentFailed: 0,
      },
    };
    return (
      pickAndAssignSlotIndependentMods(precalculatedInfo, modStatistics, items, neededStats) !==
      undefined
    );
  };

  // Answers are derived as permutations of multisets
  // e.g. for energy levels [1, 2, 1, 2, 1] we have 3 1's and 2 2's. The formula for the
  // correct number of permutations is 5!/(3!2!) = 120/(6 * 2) = 10
  // for [1, 2, 3, 1, 2] we have 5!/(2!2!1!) = 120/(2 * 2) = 30
  test.each([
    [1, 1],
    [2, 10],
    [3, 30],
    [4, 60],
    [5, 120],
  ])('generates the correct number of permutations for %i unique mods', (n, result) => {
    const mods = generalMods.map((mod, i) => modifyMod({ mod, energyVal: i % n }));
    expect(generateProcessModPermutations(mods)).toHaveLength(result);
  });

  it('can fit all mods when there are no mods', () => {
    expect(canTakeSlotIndependentMods([], [], [], items)).toBe(true);
  });

  it('can fit five general mods', () => {
    const modifiedItems = items.map((item) =>
      modifyItem({ item, energyVal: item.energy!.capacity - generalMod.energy!.val })
    );
    expect(canTakeSlotIndependentMods(generalMods, [], [], modifiedItems)).toBe(true);
  });

  test.each([0, 1, 2, 3, 4])(
    'it can fit a general mod into a single item at index %i',
    (itemIndex) => {
      const modifiedItems = items.map((item, i) =>
        modifyItem({
          item,
          energyType: generalMod.energy!.type,
          energyVal:
            itemIndex === i
              ? item.energy!.capacity - generalMod.energy!.val
              : item.energy!.capacity,
        })
      );
      expect(canTakeSlotIndependentMods([], [combatMod], [], modifiedItems)).toBe(true);
    }
  );

  test.each([
    ['can', 'combat'],
    ["can't", 'not-a-tag'],
  ])('it %s fit five combat mods', (canFit, tag) => {
    const modifiedItems = items.map((item) =>
      modifyItem({
        item,
        energyType: combatMod.energy!.type,
        energyVal: item.energy!.capacity - combatMod.energy!.val,
        compatibleModSeasons: [tag],
      })
    );
    // sanity check
    expect(canTakeSlotIndependentMods([], combatMods, [], modifiedItems)).toBe(canFit === 'can');
  });

  test.each([0, 1, 2, 3, 4])(
    'it can fit a combat mod into a single item at index %i',
    (itemIndex) => {
      const modifiedItems = items.map((item, i) =>
        modifyItem({
          item,
          energyType: combatMod.energy!.type,
          energyVal: item.energy!.capacity - combatMod.energy!.val,
          compatibleModSeasons: i === itemIndex ? [combatMod.tag!] : [],
        })
      );
      expect(canTakeSlotIndependentMods([], [combatMod], [], modifiedItems)).toBe(true);
    }
  );

  test.each([
    ['can', 'deepstonecrypt'],
    ["can't", 'not-a-tag'],
  ])('it %s fit five activity mods', (canFit, tag) => {
    const modifiedItems = items.map((item) =>
      modifyItem({
        item,
        energyType: activityMod.energy!.type,
        energyVal: item.energy!.capacity - activityMod.energy!.val,
        compatibleModSeasons: [tag],
      })
    );
    // sanity check
    expect(canTakeSlotIndependentMods([], [], activityMods, modifiedItems)).toBe(canFit === 'can');
  });

  test.each([0, 1, 2, 3, 4])(
    'it can fit a activity mod into a single item at index %i',
    (itemIndex) => {
      const modifiedItems = items.map((item, i) =>
        modifyItem({
          item,
          energyType: combatMod.energy!.type,
          energyVal: item.energy!.capacity - combatMod.energy!.val,
          compatibleModSeasons: i === itemIndex ? [activityMod.tag!] : [],
        })
      );
      expect(canTakeSlotIndependentMods([], [], [activityMod], modifiedItems)).toBe(true);
    }
  );

  it('can fit general, activity, and combat mods if there is enough energy', () => {
    const modifiedItems: ProcessItem[] = [...items];
    modifiedItems[4] = modifyItem({
      item: modifiedItems[4],
      energyType: DestinyEnergyType.Void,
      energyVal: 9,
      compatibleModSeasons: [activityMod.tag!, combatMod.tag!],
    });

    const modifiedGeneralMod = modifyMod({
      mod: generalMod,
      energyVal: 3,
    });
    const modifiedCombatMod = modifyMod({
      mod: combatMod,
      energyType: DestinyEnergyType.Void,
      energyVal: 3,
    });
    const modifiedActivityMod = modifyMod({
      mod: activityMod,
      energyType: DestinyEnergyType.Void,
      energyVal: 3,
    });

    expect(
      canTakeSlotIndependentMods(
        [modifiedGeneralMod],
        [modifiedCombatMod],
        [modifiedActivityMod],
        modifiedItems
      )
    ).toBe(false);
  });

  test.each(['general', 'combat', 'activity'])(
    "can't fit mods if %s mods have too much energy",
    (modType) => {
      const modifiedItems: ProcessItem[] = [...items];
      modifiedItems[4] = modifyItem({
        item: modifiedItems[4],
        energyType: DestinyEnergyType.Void,
        energyVal: 9,
        compatibleModSeasons: [activityMod.tag!, combatMod.tag!],
      });

      const modifiedGeneralMod = modifyMod({
        mod: generalMod,
        energyVal: modType === 'general' ? 4 : 3,
      });
      const modifiedCombatMod = modifyMod({
        mod: combatMod,
        energyType: DestinyEnergyType.Void,
        energyVal: modType === 'combat' ? 4 : 3,
      });
      const modifiedActivityMod = modifyMod({
        mod: activityMod,
        energyType: DestinyEnergyType.Void,
        energyVal: modType === 'activity' ? 4 : 3,
      });

      expect(
        canTakeSlotIndependentMods(
          [modifiedGeneralMod],
          [modifiedCombatMod],
          [modifiedActivityMod],
          modifiedItems
        )
      ).toBe(false);
    }
  );

  test.each(['general', 'combat', 'activity'])(
    "can't fit mods if a %s mod has the wrong element",
    (modType) => {
      const modifiedItems: ProcessItem[] = [...items];
      modifiedItems[4] = modifyItem({
        item: modifiedItems[4],
        energyType: DestinyEnergyType.Void,
        energyVal: 9,
        compatibleModSeasons: [activityMod.tag!, combatMod.tag!],
      });

      const modifiedGeneralMod = modifyMod({
        mod: generalMod,
        energyVal: 3,
      });
      const modifiedCombatMod = modifyMod({
        mod: combatMod,
        energyType: modType === 'combat' ? DestinyEnergyType.Arc : DestinyEnergyType.Void,
        energyVal: 3,
      });
      const modifiedActivityMod = modifyMod({
        mod: activityMod,
        energyType: modType === 'activity' ? DestinyEnergyType.Arc : DestinyEnergyType.Void,
        energyVal: 3,
      });

      expect(
        canTakeSlotIndependentMods(
          [modifiedGeneralMod],
          [modifiedCombatMod],
          [modifiedActivityMod],
          modifiedItems
        )
      ).toBe(false);
    }
  );
});
