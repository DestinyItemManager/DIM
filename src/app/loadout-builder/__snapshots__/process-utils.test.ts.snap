// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process-utils auto mods the problem is solvable 1`] = `
[
  {
    "generalModsCosts": [
      1,
    ],
    "modEnergyCost": 1,
    "modHashes": [
      1703647492,
    ],
    "numArtificeMods": 0,
    "numGeneralMods": 1,
  },
  {
    "generalModsCosts": [
      4,
    ],
    "modEnergyCost": 4,
    "modHashes": [
      4204488676,
    ],
    "numArtificeMods": 0,
    "numGeneralMods": 1,
  },
  {
    "generalModsCosts": [],
    "modEnergyCost": 0,
    "modHashes": [
      617569843,
      617569843,
      617569843,
      617569843,
    ],
    "numArtificeMods": 4,
    "numGeneralMods": 0,
  },
  {
    "generalModsCosts": [
      2,
    ],
    "modEnergyCost": 2,
    "modHashes": [
      350061697,
    ],
    "numArtificeMods": 0,
    "numGeneralMods": 1,
  },
]
`;
